src/reaper/signal.rs

```diff
@@ -0,0 +1,239 @@
+            self.pipe.wait().await;
+
+            // Notify all listeners waiting on the SIGCHLD event.
+            self.sigchld.notify(std::usize::MAX);
+
+            // Reap zombie processes, but make sure we don't hold onto the lock for too long!
+            let mut zombies = mem::take(&mut *self.zombies.lock().unwrap());
+            let mut i = 0;
+            'reap_zombies: loop {
+                for _ in 0..50 {
+                    if i >= zombies.len() {
+                        break 'reap_zombies;
+                    }
+
+                    if let Ok(None) = zombies[i].try_wait() {
+                        i += 1;
+                    } else {
+                        zombies.swap_remove(i);
+                    }
+                }
+
+                // Be a good citizen; yield if there are a lot of processes.
+                //
+                // After we yield, check if there are more zombie processes.
+                future::yield_now().await;
+                zombies.append(&mut self.zombies.lock().unwrap());
+            }
+
+            // Put zombie processes back.
+            self.zombies.lock().unwrap().append(&mut zombies);
+        }
+    }
+
+    /// Register a process with this reaper.
+    pub(crate) fn register(&'static self, child: std::process::Child) -> io::Result<ChildGuard> {
+        self.pipe.register(&child)?;
+        Ok(ChildGuard { inner: Some(child) })
+    }
+
+    /// Wait for an event to occur for a child process.
+    pub(crate) async fn status(
+        &'static self,
+        child: &Mutex<crate::ChildGuard>,
+    ) -> io::Result<std::process::ExitStatus> {
+        loop {
+            // Wait on the child process.
+            if let Some(status) = child.lock().unwrap().get_mut().try_wait()? {
+                return Ok(status);
+            }
+
+            // Start listening.
+            event_listener::listener!(self.sigchld => listener);
+
+            // Try again.
+            if let Some(status) = child.lock().unwrap().get_mut().try_wait()? {
+                return Ok(status);
+            }
+
+            // Wait on the listener.
+            listener.await;
+        }
+    }
+
+    /// Do we have any registered zombie processes?
+    pub(crate) fn has_zombies(&'static self) -> bool {
+        !self
+            .zombies
+            .lock()
+            .unwrap_or_else(|x| x.into_inner())
+            .is_empty()
+    }
+}
+
+/// The wrapper around the child.
+pub(crate) struct ChildGuard {
+    inner: Option<std::process::Child>,
+}
+
+impl ChildGuard {
+    /// Get a mutable reference to the inner child.
+    pub(crate) fn get_mut(&mut self) -> &mut std::process::Child {
+        self.inner.as_mut().unwrap()
+    }
+
+    /// Begin the reaping process for this child.
+    pub(crate) fn reap(&mut self, reaper: &'static Reaper) {
+        let mut zombies = reaper.zombies.lock().unwrap();
+        if let Ok(None) = self.get_mut().try_wait() {
```

Comment:
what if `self.inner == None`?

If that never happens, then why does `ChildGuard` contain an `Option` at all?

Reply:
why is `reaper.zombies` a Mutex of a vector, and not a channel?

Reply:
> If that never happens, then why does `ChildGuard` contain an `Option` at all?

Otherwise we are unable to `take` it out. Since `ChildGuard` is stored in an `Arc<Mutex>` this is the only option.

> why is `reaper.zombies` a Mutex of a vector, and not a channel?

A channel doesn't make much sense here, since we need a persistent list of subprocesses to poll for progress here.

Reply:
why does `zombies` get locked before the `try_wait`?

Reply:
I'm not sure, it's how it's done currently:

https://github.com/smol-rs/async-process/blob/ed17f530351227d33e3b22ecad72193da1186f37/src/lib.rs#L349-L354